src/balloonpanel/balloonpanelview.js

```diff
@@ -0,0 +1,371 @@
+
+			if ( intersectArea > maxIntersectArea ) {
+				maxIntersectRect = rect;
+				maxIntersectArea = intersectArea;
+			}
+		}
+
+		// Move balloon.
+		this.model.arrow = maxIntersectRect.name;
+		this.model.top = maxIntersectRect.top;
+		this.model.left = maxIntersectRect.left;
+	}
+}
+
+/**
+ * The balloon panel view {@link ui.Model model} interface.
+ *
+ * @interface ui.balloonPanel.BalloonPanelViewModel
+ */
+
+/**
+ * Top position of the balloon panel.
+ *
+ * @observable
+ * @member {Number} ui.balloonPanel.BalloonPanelViewModel#top
+ */
+
+/**
+ * Left position of the balloon panel.
+ *
+ * @observable
+ * @member {Number} ui.balloonPanel.BalloonPanelViewModel#left
+ */
+
+/**
+ * Max width of the balloon panel.
+ *
+ * @observable
+ * @member {Number} ui.balloonPanel.BalloonPanelViewModel#maxWidth
+ */
+
+/**
+ * Balloon panel arrow direction.
+ *
+ * @observable
+ * @member {'se'|sw'|'ne'|nw'} ui.balloonPanel.BalloonPanelViewModel#arrow
+ */
+
+/**
+ * Controls whether the balloon panel is visible or not.
+ *
+ * @observable
+ * @member {Boolean} ui.balloonPanel.BalloonPanelViewModel#isVisible
+ */
+
+/**
+ * Private helper for count and manipulate bounding box of passed element relative to whole document.
+ *
+ * **Note** It is not bounding box relative to visible viewport like result of `element.getBoundingClientRect()`. It is relative to
+ * whole document (includes scrollbars).
+ *
+ * @private
+ */
+class AbsoluteDomRect {
+	/**
+	 * Create instance of AbsoluteDomRect class.
+	 *
+	 * @param {HTMLElement|Range|Object} elementOrRangeOrRect Target object witch bounding box will be count.
+	 * @param {String} [name=undefined] Name of class instance.
+	 */
+	constructor( elementOrRangeOrRect, name ) {
+		if ( name ) {
+			this.name = name;
+		}
+
+		Object.assign( this, getAbsoluteBoundingBoxOf( elementOrRangeOrRect ) );
+	}
+
+	/**
+	 * Clone instance of this class.
+	 *
+	 * @param {String} newName Name of new instance.
+	 * @returns {AbsoluteDomRect}
+	 */
+	clone( newName ) {
+		return new AbsoluteDomRect( this, newName );
+	}
+
+	/**
+	 * Move current bounding to passed coordinates.
+	 *
+	 * @param {Number} top New to position.
+	 * @param {Number} left New left position.
+	 * @returns {AbsoluteDomRect}
+	 */
+	moveTo( { top, left } ) {
+		this.top = top;
+		this.right = left + this.width;
+		this.bottom = top + this.height;
+		this.left = left;
+
+		return this;
+	}
+
+	/**
+	 * Get intersect surface area of this AbsoluteDomRect and other AbsoluteDomRect.
+	 *
+	 * @param {AbsoluteDomRect} rect
+	 * @returns {Number} Overlap surface area.
+	 */
+	getIntersectArea( rect ) {
+		const hOverlap = Math.max( 0, Math.min( this.right, rect.right ) - Math.max( this.left, rect.left ) );
+		const vOverlap = Math.max( 0, Math.min( this.bottom, rect.bottom ) - Math.max( this.top, rect.top ) );
+
+		return hOverlap * vOverlap;
+	}
+}
+
+/**
+ * Get relative to whole document bounding box of passed HTMLElement, Range, or Rect.
```

Comment:
"Get the bounding box of `HTMLElement`, `Range`, or rect relative to the entire document (like it had `position: absolute` in CSS)."